practice/completer/site.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+/**
+ *  The Completer Constructor holds valid words that can be called through
+ * the prototype method complete, to find valid words that start with given
+ * prefix.
+ * There is also a weighting system, where everytime selectCompletion is called
+ * the selected string has weight added to it. The complete function returns
+ * an array ordered in more weighted words first.
+ */
+function Completer() {
+  this.validCompletionToWeight = {};
+}
+
+var completerProto = {
+  addCompletion: function(str) {
+    this.validCompletionToWeight[str] = 0;
+  },
+  removeCompletion: function(str) {
+    delete this.validCompletionToWeight[str];
+  },
+  complete: function(prefix) {
+    var potentialWords = _.filter(
+      _.keys(this.validCompletionToWeight), function(o) {
+        return new RegExp('^' + prefix).test(o);
+      });
```

Comment:
There's are two subtle advantages to moving it out:

1. This has to create a new regular expression object for each valid completion. If there are lots, this unnecessarily slows down the program.
1. You can then use the fact that your regexp.test() is already the function you want and can avoid needing to make a new anonymous function.

```js
var prefixPattern = new RegExp('^' + prefix);
var potentialWords = _.filter(_.keys(this.validCompletionToWeight), prefixPattern.test);
```